Refactor `GccrsArgs` structure

CohenArthur opened the pull request:
This PR cleans up the `GccrsArgs` type by splitting it in multiple modules. As pointed out by @philberty and @flip1995, the type should be doing much less than what is currently implemented.

Closes #23

CohenArthur committed:
options: Move rustc options parser to separate file

CohenArthur committed:
args: Add separate module for environment args

flip1995 commented on src/gccrs/rustc_options.rs:
```diff
@@ -0,0 +1,43 @@
+        // Parse arguments, skipping `cargo-gccrs` and `rustc` in the invocation
+        let matches = self.options.parse(&args[2..])?;
+
+        Ok(matches)
```
```suggestion
        Ok(self.options.parse(&args[2..])?)
```

And see if Clippy complains about this and tells you to remove the `Ok(_?)` wrap.

CohenArthur replied:
No complaints from clippy!

flip1995 commented on src/gccrs/args.rs:
```diff
@@ -173,10 +123,7 @@ impl GccrsArgs {
 
     /// Get the corresponding `gccrs` argument from a given `rustc` argument
     pub fn from_rustc_args(rustc_args: &[String]) -> Result<Vec<GccrsArgs>> {
```
To rustify this, you would want to implement 
```rs
impl TryFrom<RustcOptions> for Vec<GccrsArgs> { ... }
```

But this isn't possible, since you can't implement an external trait for an external type `Vec`.

What you normally do in this case is to make a wrapper type that wraps the `Vec<_>` and implements `Deref`.

For this, a possible name for the wrapper type would be `GccrsArgss` (2 `s`). And this is the point where it doesn't make sense to me why `GccrsArgs` is called `GccrsArgs`, when you need multiple instances to get all `Args` for `Gccrs`. The name "`GccrsArgs`" already implies, that you have every arg, but for some reason there is still the need for a `GccrsArgs"s"`.

From a user perspective I would expect something like this:

```rs
// Parse the arguments cargo provides for rustc and try to convert them to arguments that can be passed to gccrs
let args: GccrsArgs = RustcOptions::new().parse(rustc_args)?.try_into()?;
// build the gccrs command with the arguments converted from the rustc arguments
let gccrs = Command::new("gccrs").args(args.into());
// Run gccrs
```

Also why is `RustcOptions` called `Options` and `GccrsArgs` called `Args`?

CohenArthur replied:
Fixed in 4aa2cab and 7eb1f3b

CohenArthur replied:
Sorry about the clock skew

flip1995 commented on src/gccrs/args.rs:
```diff
@@ -94,75 +86,33 @@ pub struct GccrsArgs {
-        }
+    /// Set the callback of an argument set
+    pub fn set_callback(&mut self, function: &'static dyn Fn(&GccrsArgs) -> Result) {
+        self.callback = Some(function);
```
I still think, that `GccrsArgs` should handle nothing but the parsing/conversion of arguments.

CohenArthur replied:
Fixed in 8b442e6

CohenArthur committed:
rustc_options: Parse --cfg and --cap-lints

[--cap-lints](https://github.com/Rust-GCC/gccrs/issues/575)
[--cfg](https://github.com/Rust-GCC/gccrs/issues/574)

flip1995 reviewed (changes requested):
Oops, forgot to submit my review last Friday...

CohenArthur committed:
args: Wrap vector of GccrsArgs in GccrsArgsCollection structure

Co-authored-by: flip1995 <hello@philkrones.com>

CohenArthur committed:
gccrs: Rename structure to benefit from namespacing

Co-authored-by: flip1995 <hello@philkrones.com>

CohenArthur committed:
rustc_options: Format parse() method better

Co-authored-by: flip1995 <hello@philkrones.com>

CohenArthur committed:
args: Remove callback from GccrsArgs structure

Co-authored-by: flip1995 <hello@philkrones.com>

flip1995 commented on src/gccrs/args.rs:
```diff
@@ -1,11 +1,85 @@
 
-use getopts::{Matches, Options};
+    fn try_from(rustc_args: &[String]) -> Result<OptionsCollection> {
+        let matches = RustcOptions::new().parse(rustc_args)?;
```
I would implement this on `RustcOptions` and not on `&[String]` directly.

CohenArthur replied:
Fixed in 2b39f21

flip1995 commented on src/gccrs/args.rs:
```diff
@@ -1,11 +1,85 @@
+/// Therefore, we need to wrap an unknown amount of sets of `gccrs` arguments in order to
+/// mimic a single `rustc` invokation. Later on, we need to iterate over those sets and
+/// spawn a new `gccrs` command for each of them.
+pub struct OptionsCollection {
+    args_set: Vec<Options>,
+}
```
IIUC most of the `Options` are common to all `gccrs` invocations. So

```bash
rustc --crate-type=dyn --crate-type=bin src/<file>.rs --other-flag-1 --other-flag-2
```

becomes

```bash
gccrs src/<file>.rs --converted-other-flag-1 --converted-other-flag-2  # bin
gccrs -shared src/<file>.rs -o libfile.so --converted-other-flag-1 --converted-other-flag-2  # shared lib
```

where `--converted-other-flag-1 --converted-other-flag-2` are common to all invocations. So why not only store them once?

```rust
pub struct OptionsCollection {
    common_args: Vec<Arg>,
    crate_types: Vec<CrateType>,
    ... // Other fields that require multiple gccrs invocations, if any
}
```

I think that would make things way cleaner and would avoid saving the common options and input+output file name `N` times. 

Maybe even:

```rust
pub struct Args {
    input_files: Vec<String/PathBuf>,
    output_file: String/PathBuf,
    common_args: Vec<Arg>,
    crate_types: Vec<CrateType>,
    ... // Other fields that require multiple gccrs invocations, if any
}
```

flip1995 replied:
And only once you generate the `gccrs` cmmands, you will build the whole `Vec<Vec<String>>` argument matrix.

CohenArthur replied:
As of right now, there aren't many shared options since the translator does not do much. I think your design makes a lot of sense for when we will handle all Rustc arguments (for example, after #46 is done). The only shared arguments for now are `-fPIE` and `-pie` on the `main` branch. I'm not against it but it's a bit hard to wrap my head around it with the current implementation. How do you feel about opening an issue and doing this in a later PR? If you think it's really important I'll try and do it now :)

flip1995 replied:
IMO simply the fact that you currently clone the input and output names N times would already warrant this design. And in the future there probably will be more shared arguments. So the current design, where you have to clone all of those arguments N times isn't scalable. 

Since this PR already improves the situation a bit, I'm good with merging this for now and to keep the technical debt.

flip1995 reviewed (commented):
I may have an idea on how to better manage the GccrsArgs.

CohenArthur committed:
args: Implement TryFrom<RustcArgs> for ArgsCollection

Co-authored-by: flip1995 <hello@philkrones.com>

CohenArthur committed:
args: Handle UTF8 errors more gracefully

CohenArthur commented:
The tests pass on my machine, even on a clean clone of the repo, but not on the CI. I'm going to do some `tmate` debugging

CohenArthur commented:
Tests are passing on tmate when running the docker image and cloning the repo too...

CohenArthur commented:
Found the bug, fixing it. No idea why it works on the `tmate` session however

CohenArthur committed:
object_file_name: Set extension instead of adding an extra extension to
object file name

The previous implementation would try to add a new component to the
cloned PathBuf, which would act as if the original output file was a
directory. This is not the correct behavior. Instead, set a different
extension to the cloned PathBuf so that `lib<name>.a` becomes
`lib<name>.tmp.o`, which is basically the same
